From a0311bd9846186b1de4de19b5fd4ee192eebed57 Mon Sep 17 00:00:00 2001
From: "freedom.tan" <freedom.tan@mediatek.com>
Date: Thu, 8 Oct 2015 14:41:52 +0800
Subject: [PATCH 58/58] 8173 power model measured by servo board

Provide power/engegy model measured by servo board.
Fix the order of ca53 and ca57/ca72 clusters.
---
 arch/arm64/boot/dts/mediatek/mt8173.dtsi |   4 +-
 arch/arm64/kernel/energy_model.h         | 107 +++++++++++++++++++++++++++++++
 2 files changed, 109 insertions(+), 2 deletions(-)

diff --git a/arch/arm64/boot/dts/mediatek/mt8173.dtsi b/arch/arm64/boot/dts/mediatek/mt8173.dtsi
index 1b1fd9b..dd7de00 100644
--- a/arch/arm64/boot/dts/mediatek/mt8173.dtsi
+++ b/arch/arm64/boot/dts/mediatek/mt8173.dtsi
@@ -165,7 +165,7 @@
 			#cooling-min-level = <0>;
 			#cooling-max-level = <7>;
 			dynamic-power-coefficient = <263>;
-			capacity-dmips-mhz = <485>;
+			capacity-dmips-mhz = <573>;
 		};
 
 		cpu1: cpu@1 {
@@ -183,7 +183,7 @@
 			#cooling-cells = <2>;
 			#cooling-min-level = <0>;
 			#cooling-max-level = <7>;
-			capacity-dmips-mhz = <485>;
+			capacity-dmips-mhz = <573>;
 		};
 
 		cpu2: cpu@100 {
diff --git a/arch/arm64/kernel/energy_model.h b/arch/arm64/kernel/energy_model.h
index 0a7c134..e25c247 100644
--- a/arch/arm64/kernel/energy_model.h
+++ b/arch/arm64/kernel/energy_model.h
@@ -107,6 +107,107 @@ static struct sched_group_energy energy_core_juno_a57 = {
 	  .cap_states     = cap_states_core_juno_a57,
 };
 
+/* MT8173 */
+static struct idle_state idle_states_cluster_mt8173_a53[] = {
+	{ .power = 184 }, /* arch_cpu_idle() (active idle) = WFI */
+	{ .power = 184 }, /* WFI */
+	{ .power = 147 }, /* cpu-sleep-0 */
+	{ .power = 4 }, /* cluster-sleep-0 */
+};
+
+static struct idle_state idle_states_cluster_mt8173_a57[] = {
+	{ .power = 171 }, /* arch_cpu_idle() (active idle) = WFI */
+	{ .power = 171 }, /* WFI */
+	{ .power = 100 }, /* cpu-sleep-0 */
+	{ .power = 18 }, /* cluster-sleep-0 */
+};
+
+static struct capacity_state cap_states_cluster_mt8173_a53[] = {
+	{ .cap =  184, .power = 174, },
+	{ .cap =  256, .power = 107, },
+	{ .cap =  368, .power = 138, },
+	{ .cap =  399, .power = 145, },
+	{ .cap =  430, .power = 144, },
+	{ .cap =  512, .power = 150, },
+	{ .cap =  552, .power = 165, },
+	{ .cap =  573, .power = 164, },
+};
+
+static struct capacity_state cap_states_cluster_mt8173_a57[] = {
+	{ .cap =  256, .power = 56, },
+	{ .cap =  358, .power = 89, },
+	{ .cap =  512, .power = 78, },
+	{ .cap =  614, .power = 47, },
+	{ .cap =  716, .power = 101, },
+	{ .cap =  819, .power = 122, },
+	{ .cap =  921, .power = 103, },
+	{ .cap =  1024, .power = 150, },
+};
+
+static struct sched_group_energy energy_cluster_mt8173_a53 = {
+	.nr_idle_states = ARRAY_SIZE(idle_states_cluster_mt8173_a53),
+	.idle_states    = idle_states_cluster_mt8173_a53,
+	.nr_cap_states  = ARRAY_SIZE(cap_states_cluster_mt8173_a53),
+	.cap_states     = cap_states_cluster_mt8173_a53,
+};
+
+static struct sched_group_energy energy_cluster_mt8173_a57 = {
+	.nr_idle_states = ARRAY_SIZE(idle_states_cluster_mt8173_a57),
+	.idle_states    = idle_states_cluster_mt8173_a57,
+	.nr_cap_states  = ARRAY_SIZE(cap_states_cluster_mt8173_a57),
+	.cap_states     = cap_states_cluster_mt8173_a57,
+};
+
+static struct idle_state idle_states_core_mt8173_a53[] = {
+	{ .power = 6 }, /* arch_cpu_idle() (active idle) = WFI */
+	{ .power = 6 }, /* WFI */
+	{ .power = 0 }, /* cpu-sleep-0 */
+	{ .power = 0 }, /* cluster-sleep-0 */
+};
+
+static struct idle_state idle_states_core_mt8173_a57[] = {
+	{ .power = 15 }, /* arch_cpu_idle() (active idle) = WFI */
+	{ .power = 15 }, /* WFI */
+	{ .power = 0  }, /* cpu-sleep-0 */
+	{ .power = 0  }, /* cluster-sleep-0 */
+};
+
+static struct capacity_state cap_states_core_mt8173_a53[] = {
+	{ .cap =  184, .power = 57, },
+	{ .cap =  256, .power = 156, },
+	{ .cap =  368, .power = 255, },
+	{ .cap =  399, .power = 299, },
+	{ .cap =  430, .power = 339, },
+	{ .cap =  512, .power = 472, },
+	{ .cap =  552, .power = 529, },
+	{ .cap =  573, .power = 574, },
+};
+
+static struct capacity_state cap_states_core_mt8173_a57[] = {
+	{ .cap =  256, .power = 316, },
+	{ .cap =  358, .power = 374, },
+	{ .cap =  512, .power = 640, },
+	{ .cap =  614, .power = 885, },
+	{ .cap =  716, .power = 1079, },
+	{ .cap =  819, .power = 1376, },
+	{ .cap =  921, .power = 1738, },
+	{ .cap =  1024, .power = 2207, },
+};
+
+static struct sched_group_energy energy_core_mt8173_a53 = {
+	.nr_idle_states = ARRAY_SIZE(idle_states_core_mt8173_a53),
+	.idle_states    = idle_states_core_mt8173_a53,
+	.nr_cap_states  = ARRAY_SIZE(cap_states_core_mt8173_a53),
+	.cap_states     = cap_states_core_mt8173_a53,
+};
+
+static struct sched_group_energy energy_core_mt8173_a57 = {
+	.nr_idle_states = ARRAY_SIZE(idle_states_core_mt8173_a57),
+	.idle_states    = idle_states_core_mt8173_a57,
+	.nr_cap_states  = ARRAY_SIZE(cap_states_core_mt8173_a57),
+	.cap_states     = cap_states_core_mt8173_a57,
+};
+
 /* HiKey */
 
 static struct idle_state idle_states_system_hikey[] = {
@@ -194,9 +295,15 @@ static struct energy_model hikey_model = {
 	{ &energy_system_hikey, &energy_system_hikey, },
 };
 
+static struct energy_model mt8173_model = {
+	{ &energy_core_mt8173_a53, &energy_core_mt8173_a57, },
+	{ &energy_cluster_mt8173_a53, &energy_cluster_mt8173_a57, },
+	{},
+};
 static struct of_device_id model_matches[] = {
 	{ .compatible = "arm,juno", .data = &juno_model },
 	{ .compatible = "hisilicon,hi6220-hikey", .data = &hikey_model },
+	{ .compatible = "mediatek,mt8173", .data = &mt8173_model },
 	{},
 };
 
-- 
2.7.4

